### DISCUSSION TITLE
Using `targets` for workflows with many side-effects

### DISCUSSION THREAD

**Discussion by Aariq** [ropensci/targets#903] [Help]
I'm working on an already existing mess of a project that has many R functions that are called for their side-effects like writing .Rdata files with multiple R objects that get loaded in later steps, dumping objects into the global environment (😠).  It doesn't "feel" very `targets` friendly so I've been wary of trying to use `targets` for this project, although I'm sure it would be helpful if I got it working.  Anyone have any advice or examples that might be helpful/inspiring?

**Comment by stuvet** [comment]
I'd bet that's a great case for targets! Sounds like you could cut out all the .RData dumps completely & simplify the code a heap? You might be surprised how small a pipeline you end up with if a lot of the functions are I/O, checks & validation. Far more scalable too.

Tip: I use static branching via `tarchetypes::tar_eval` a lot to specify branch names for big pipelines - it's a bit more work but knowing exactly what each target is named (rather than having the random names of dynamic branching) is a *huge* win for development & maintenance of big pipelines. I probably wasted too long trying to 'keep it simple' with dynamic branching & ended up getting frustrated until I forced myself to learn static - need to get friendly with `rlang::syms`, & use `tar_manifest` to troubleshoot the code you've produced. It's *far* more powerful & near-infinitely flexible.

Motivation: Targets has changed my life. Just do it! 😁

**Reply by Aariq** [reply]
Yeah, you're right that pieces of this workflow are basically trying to do (poorly) what targets does.  But the code that is dumping stuff into .Rdata files are functions in R packages.  So I'm kindof constrained by having to work around this existing awful behavior unless I really fight to re-design a lot of how this package ecosystem works at a high level (which is not what I'm getting paid for).

It still might be worth a try...

**Reply by stuvet** [reply]
I guess
@wlandau addressed a similar problem was addressed with `stantargets` - the underlying `cmdstanr` package only holds R6 objects in memory, but their methods compile c++ code, dump flat input files to disk & query flat output files after running the binary.

Not necessarily a natural problem to solve with `targets` but it'll be well worth it in the end. Except for the "not what I'm paid for" bit - maybe life is too short in this case?

**Reply by Aariq** [reply]
Maybe I'll poke around in the code for `stantargets` for some inspiration.  Thanks!

**Reply by wlandau** [reply]
Yeah, I like that approach. On your end, you could write a wrapper around the package function that temporarily isolates the side effect and return an R object. Sketch:

```r
pure_wrapper_function <- function(...) {

  # Change to a unique temporary directory
  # so .Rdata does not pollute the workspace
  # or clash with another target's .Rdata:
  dir <- tempfile()
  fs::dir_create(dir)
  withr::local_dir(dir)

  # Optional: clean up .Rdata when you are done,
  # if a lot of data in /tmp is going to be a problem:
  on.exit(unlink(".RData"), add = TRUE)

  # Write .Rdata
  package::function_that_writes_rdata()

  # Load R data into a local environment:
  envir <- new.env(parent = emptyenv())
  load(".Rdata", envir = envir)

  # If you know what data objects to look for,
  # find them in the environment and return them.
  list(object1 = envir$object1, object2 = envir$object2) 
}
```

You may not need them, but [`tarchetypes` hooks](https://docs.ropensci.org/tarchetypes/reference/index.html#hooks) help apply these kinds of wrappers to many targets at once.

**Reply by Aariq** [reply]
Thanks! This looks like it'll be a super helpful example if I decide to go this route.
